# Task 1: Understanding Dataset & Data Types

**Tools:**
* Python (Pandas, NumPy)
* Jupyter Notebook / Google Colab

**Datasets:**
1.  **Titanic Dataset** (Classification)
2.  **Students Performance Dataset** (Regression/Analysis)

---

## 📋 Objectives & Hints

1.  **Load the dataset** using Pandas and display the first and last few records to understand the structure of rows and columns.
2.  **Identify feature types** (numerical, categorical, ordinal, and binary) manually by inspecting column names and values.
3.  **Inspect data** using `df.info()` and `df.describe()` to understand data types, null values, and statistical summaries.
4.  **Analyze distribution** by checking unique values in categorical columns.
5.  **Determine ML Suitability** by identifying the target variable and input features.
6.  **Assess Dataset Size** to determine if it is suitable for machine learning.
7.  **Quality Check** by writing clear observations about issues like missing values or class imbalance.

---

## 🚀 Deliverables
* Jupyter Notebook with code implementation.
* Dataset analysis report (Markdown cell).
* **Outcome:** Understanding data structure, types, and ML readiness.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns # Used here to easily load the Titanic dataset

# 1. Load Titanic Dataset (via Seaborn library)
df_titanic = sns.load_dataset('titanic')

print("Titanic Dataset Loaded Successfully")

Titanic Dataset Loaded Successfully


In [2]:
# Display first and last 5 rows
print("--- First 5 Rows ---")
display(df_titanic.head())

print("\n--- Last 5 Rows ---")
display(df_titanic.tail())

# Check Dataset Size
print(f"\nShape of Dataset (Rows, Columns): {df_titanic.shape}")

--- First 5 Rows ---


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True


--- Last 5 Rows ---


survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True


Shape of Dataset (Rows, Columns): (891, 15)


In [3]:
# Check Data Types and Null Values
print("--- Info & Missing Values ---")
df_titanic.info()

# Statistical Summary for Numerical Columns
print("\n--- Numerical Summary ---")
display(df_titanic.describe())

# Statistical Summary for Categorical Columns (include='O')
print("\n--- Categorical Summary ---")
display(df_titanic.describe(include=['O', 'category']))

--- Info & Missing Values ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB

--- Numerical Summary ---


survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


--- Categorical Summary ---


sex embarked  class  who deck  embark_town alive
count    891      889    891  891  203          889   891
unique     2        3      3    3    7            3     2
top     male        S  Third  man    C  Southampton    no
freq     577      644    491  537   59          644   549

In [4]:
# Check unique values in a specific categorical column
print("Unique Embark Towns:", df_titanic['embark_town'].unique())
print("Unique Classes:", df_titanic['class'].unique())

# Check for Data Imbalance in Target Variable
print("\n--- Target Variable Distribution (Survived) ---")
print(df_titanic['survived'].value_counts(normalize=True) * 100)

Unique Embark Towns: ['Southampton' 'Cherbourg' 'Queenstown' nan]
Unique Classes: ['Third', 'First', 'Second']
Categories (3, object): ['First', 'Second', 'Third']

--- Target Variable Distribution (Survived) ---
survived
0    61.616162
1    38.383838
Name: proportion, dtype: float64


## Project: Titanic Dataset Exploratory Analysis

## 1. Data Structure Overview
* **Dataset Size:** The dataset contains **891 rows** and **15 columns**.
* **Suitability:** This size is considered small but sufficient for learning basic Machine Learning (ML) classification algorithms.

## 2. Variable Identification
Based on `df.head()` and manual inspection, the features can be classified as:

* **Numerical:** `age`, `fare`, `sibsp` (Siblings/Spouses), `parch` (Parents/Children).
* **Categorical (Nominal):** `sex`, `embarked`, `who`, `deck`.
* **Categorical (Ordinal):** `class` (First > Second > Third), `pclass`.
* **Binary:** `adult_male` (True/False), `alone` (True/False).
* **Target Variable:** `survived` (0 = No, 1 = Yes). This makes it a **Classification** problem.

## 3. Data Quality & Observations

### Missing Values:
* The `deck` column has a massive number of missing values (approx **77%**). It may need to be dropped.
* The `age` column has roughly **20%** missing values. We will need to fill these (imputation) before ML modeling.

### Imbalance:
* The target variable `survived` shows that roughly **61% died** and **38% survived**. This is a slight imbalance but acceptable for standard training without heavy modification.

## 4. ML Readiness Conclusion
The dataset is **Partially Ready**.

* **Action Required:** We must handle the missing `age` data and convert categorical text data (like 'male'/'female') into numbers (0/1) before feeding it into a model.